src/pybullet_blocks/envs/cleanup_table_env.py

```diff
@@ -780,6 +845,154 @@ def is_toy_in_bin(self, toy_id: int) -> bool:
+        elif action_obj.gripper_action == -1:
+            # Check if any object is close enough to the end effector position
+            # and grasp if so.
+            for object_id in self._get_movable_object_ids():
+                if self._is_graspable(object_id):
+                    world_to_robot = self.robot.get_end_effector_pose()
+                    world_to_object = get_pose(object_id, self.physics_client_id)
+                    self.current_grasp_transform = multiply_poses(
+                        world_to_robot.invert(), world_to_object
+                    )
+                    self.current_held_object_id = object_id
+                    break
+
+        # Store original state.
+        original_joints = self.robot.get_joint_positions()
+
+        # Update the robot position and held objects.
+        clipped_joints = np.clip(
+            joint_arr, self.robot.joint_lower_limits, self.robot.joint_upper_limits
+        )
+        self._set_robot_and_held_object_state(clipped_joints.tolist())
+
+        has_collision = False
+
+        # Check robot-table penetration.
+        thresh = self.scene_description.robot_table_penetration_dist
+        if check_body_collisions(
+            self.robot.robot_id,
+            self.table_id,
+            self.physics_client_id,
+            distance_threshold=-thresh,
+        ):
+            has_collision = True
+
+        # Check held object-table penetration.
+        if self.current_held_object_id is not None:
+            thresh = self.scene_description.grasped_object_table_penetration_dist
+            if check_body_collisions(
+                self.current_held_object_id,
+                self.table_id,
+                self.physics_client_id,
+                distance_threshold=-thresh,
+            ):
+                has_collision = True
+
+        # If collision detected, revert to original state and return.
+        if has_collision:
+            self._set_robot_and_held_object_state(original_joints)
+            observation = self.get_state().to_observation()
+            penetration_penalty = self.scene_description.penetration_penalty
+            reward = -1 * penetration_penalty
+            return observation, reward, False, False, self._get_info()
+
+        # Run physics simulation.
+        for _ in range(self._num_sim_steps_per_step):
+            p.stepSimulation(physicsClientId=self.physics_client_id)
+
+        # Reset the robot and held object again after physics to ensure that position
+        # control is exact.
+        self._set_robot_and_held_object_state(clipped_joints.tolist())
+
+        # Check goal.
+        terminated = self._get_terminated()
+        reward = self._get_reward()
+        truncated = False
+        observation = self.get_state().to_observation()
+        info = self._get_info()
+        self._timestep += 1
+
+        return observation, reward, terminated, truncated, info
+
+    def _is_graspable(self, object_id: int) -> bool:
```

Comment:
are `is_object_ready_pick` and `_is_graspable` redundant? can you help me understand the relationship?